src/main.cpp

```diff
@@ -450,32 +457,82 @@ Napi::Value DDWAFContext::run(const Napi::CallbackInfo& info) {
     case DDWAF_ERR_INVALID_OBJECT:
     case DDWAF_ERR_INVALID_ARGUMENT:
       res.Set("errorCode", Napi::Number::New(env, code));
-      ddwaf_result_free(&result);
+      ddwaf_object_free(&result);
       return res;
     default:
       break;
   }
-  // there is no error. We need to collect perf data
+
+  // No error. Collect result data and return
+
+  const ddwaf_object *events = nullptr, *actions = nullptr, *attributes = nullptr,
+                     *keep = nullptr, *duration = nullptr, *run_timeout = nullptr;
+
+  for (size_t i = 0; i < ddwaf_object_size(&result); ++i) {
+    const ddwaf_object *child = ddwaf_object_get_index(&result, i);
+    if (child == nullptr) {
+      mlog("ddwaf result child is null")
+      continue;
+    }
+
+    size_t length = 0;
+    const char *key = ddwaf_object_get_key(child, &length);
+    if (key == nullptr) {
+      mlog("ddwaf result key is null")
+      continue;
+    }
+
+    if (length == EVENTS_LEN && memcmp(key, "events", EVENTS_LEN) == 0) {
+      events = child;
+    } else if (length == ACTIONS_LEN && memcmp(key, "actions", ACTIONS_LEN) == 0) {
+      actions = child;
+    } else if (length == ATTRIBUTES_LEN && memcmp(key, "attributes", ATTRIBUTES_LEN) == 0) {
+      attributes = child;
+    } else if (length == KEEP_LEN && memcmp(key, "keep", KEEP_LEN) == 0) {
+      keep = child;
+    } else if (length == DURATION_LEN && memcmp(key, "duration", DURATION_LEN) == 0) {
+      duration = child;
+    } else if (length == TIMEOUT_LEN && memcmp(key, "timeout", TIMEOUT_LEN) == 0) {
+      run_timeout = child;
+    }
+  }
 
   mlog("Set timeout");
-  res.Set("timeout", Napi::Boolean::New(env, result.timeout));
+  if (run_timeout && run_timeout->type == DDWAF_OBJ_BOOL) {
```

Comment:
why not do the type checking directly in the loop above ?

Reply:
well by saving the condition inside the loop, you save on the `events = child;` it's not much tho, so if you prefere it like this, i'm fine with it